implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,611 @@
+  defp identity_exchange(%Channel{role: :responder} = state, h, encrypt_st, decrypt_st) do
+    {:ok, encrypt_st} = send_identity_proof(IdentityHandshake.Request, state, h, encrypt_st)
+    channel_state = identity_exchange_state(h, encrypt_st, decrypt_st, state.channel_state)
+    {:ok, %Channel{state | channel_state: channel_state}}
   end
 
-  defp notify_callback_route(_data) do
-    :ok
+  defp identity_exchange_state(h, encrypt_st, decrypt_st, %Handshaking{waiting: w, timer: t}) do
+    %IdentityExchange{h: h, encrypt_st: encrypt_st, decrypt_st: decrypt_st, waiting: w, timer: t}
   end
 
-  # application facing address is plaintext address
-  defp setup_plaintext_address(options, data) do
-    case Keyword.get(options, :address) do
-      nil -> {:error, {:option_is_nil, :address}}
-      plaintext_address -> {:ok, Map.put(data, :plaintext_address, plaintext_address)}
-    end
+  defp send_identity_proof(type, state, h, encrypt_st) do
+    contact_data = Identity.get_data(state.identity)
+    {:ok, proof} = Identity.create_signature(state.identity, h, state.vault_name)
+    identity_msg = struct(type, %{contact: contact_data, proof: proof})
+    encoded_identity_proof = IdentityHandshake.encode(identity_msg)
+
+    msg = %Message{
+      payload: encoded_identity_proof,
+      onward_route: state.peer_identity_route,
+      return_route: [state.inner_address]
+    }
+
+    send_over_encrypted_channel(msg, encrypt_st, state.peer_route, state.inner_address)
   end
 
-  # network facing address is ciphertext address
-  defp setup_ciphertext_address(options, data) do
-    ## TODO: use a different prefix?
-    address_prefix = Keyword.get(options, :address_prefix, "")
-    ciphertext_address = Node.get_random_unregistered_address(address_prefix)
+  defp handle_inner_message_impl(message, %Channel{channel_state: %Handshaking{xx: xx}} = state) do
+    with {:ok, data, ""} <- :bare.decode(message.payload, :data),
+         {:ok, next} <- XX.in_payload(xx, data) do
+      continue_handshake(next, %Channel{state | peer_route: message.return_route})
+    end
+  end
 
-    with :ok <- Node.register_address(ciphertext_address, __MODULE__) do
-      {:ok, Map.put(data, :ciphertext_address, ciphertext_address)}
+  defp handle_inner_message_impl(message, %Channel{channel_state: channel_state} = state) do
+    with {:ok, ciphertext, ""} <- :bare.decode(message.payload, :data),
+         {:ok, plaintext, decrypt_st} <-
+           Decryptor.decrypt("", ciphertext, channel_state.decrypt_st) do
+      case Wire.decode(plaintext, :secure_channel) do
+        {:ok, message} ->
+          handle_decrypted_message(message, %Channel{
+            state
+            | channel_state: %{channel_state | decrypt_st: decrypt_st}
+          })
+
+        {:error, reason} ->
+          {:error, reason}
+      end
+    else
+      # The message couldn't be decrypted.  State remains unchanged
+      _error -> {:ok, state}
```

Comment:
Should this be logged, metered with telemetry, or anything similar to give us some insight later?

Reply:
added it to logs